File: backend/models/weekmenu.py
from sqlalchemy import Column, Integer, Date, ForeignKey, Boolean
from sqlalchemy.orm import relationship
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import date
from models.product import Base
from models.recipe import RecipeDB

class WeekMenuDB(Base):
    __tablename__ = "week_menus"
    
    id = Column(Integer, primary_key=True, index=True)
    start_date = Column(Date, nullable=False, index=True)
    end_date = Column(Date, nullable=False, index=True)
    
    # Relationship to menu days
    days = relationship("MenuDayDB", back_populates="week_menu", cascade="all, delete-orphan")

class MenuDayDB(Base):
    __tablename__ = "menu_days"
    
    id = Column(Integer, primary_key=True, index=True)
    week_menu_id = Column(Integer, ForeignKey("week_menus.id"), nullable=False)
    date = Column(Date, nullable=False)
    recipe_id = Column(Integer, ForeignKey("recipes.id"), nullable=True)
    servings = Column(Integer, nullable=False)
    add_to_shopping_list = Column(Boolean, nullable=False, default=True)
    
    # Relationships
    week_menu = relationship("WeekMenuDB", back_populates="days")
    recipe = relationship("RecipeDB")

# Pydantic schemas
class MenuDayBase(BaseModel):
    date: date
    recipe_id: Optional[int] = None
    servings: Optional[int] = None
    add_to_shopping_list: bool = True

class MenuDayCreate(MenuDayBase):
    pass

class MenuDayResponse(MenuDayBase):
    id: int
    recipe: Optional[dict] = None  # Basic recipe info
    
    class Config:
        from_attributes = True

class WeekMenuBase(BaseModel):
    start_date: date = Field(..., description="Start date of the week")
    end_date: date = Field(..., description="End date of the week")

class WeekMenuCreate(WeekMenuBase):
    days: List[MenuDayCreate] = []

class WeekMenuResponse(WeekMenuBase):
    id: int
    days: List[MenuDayResponse] = []
    
    class Config:
        from_attributes = True

File: frontend/src/models/WeekMenu.js
export class WeekMenu {
    constructor(data = {}) {
      this.id = data.id || null;
      this.start_date = data.start_date || '';
      this.end_date = data.end_date || '';
      this.days = data.days ? data.days.map(day => new MenuDay(day)) : [];
    }
  
    static fromAPI(apiData) {
      return new WeekMenu(apiData);
    }
  
    toAPI() {
      return {
        start_date: this.start_date,
        end_date: this.end_date,
        days: this.days.map(day => day.toAPI())
      };
    }
  
    addDay(dayData = {}) {
      this.days.push(new MenuDay(dayData));
    }
  
    removeDay(index) {
      this.days.splice(index, 1);
    }
  
    getDayByDate(dateStr) {
      return this.days.find(day => day.date === dateStr);
    }
  }
  
  export class MenuDay {
    constructor(data = {}) {
      this.id = data.id || null;
      this.date = data.date || '';
      this.recipe_id = data.recipe_id || null;
      this.recipe = data.recipe || null;
      this.servings = data.servings || null;
      this.add_to_shopping_list = data.add_to_shopping_list !== undefined ? data.add_to_shopping_list : true;
    }
  
    static fromAPI(apiData) {
      return new MenuDay(apiData);
    }
  
    toAPI() {
      return {
        date: this.date,
        recipe_id: this.recipe_id,
        servings: this.servings,
        add_to_shopping_list: this.add_to_shopping_list
      };
    }
  }